From 3d87a3fe565d921b8d6296b105e08afeb02d798f Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Wed, 30 Nov 2022 12:51:08 +0530
Subject: [PATCH] cfg80211: get link information of associated station

struct station_info contains all information related to
station dump.

Currently, no information present in struct
station_info related to link of associated station.

So, add link related information for associated station.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 include/net/cfg80211.h |  7 +++++++
 net/wireless/nl80211.c | 18 ++++++++++++++++++
 2 files changed, 25 insertions(+)

--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -1966,6 +1966,11 @@ struct cfg80211_tid_stats {
 
 #define IEEE80211_MAX_CHAINS	4
 
+struct link_station_info {
+	u8 addr[ETH_ALEN] __aligned(2);
+	/* TODO: packet stats */
+};
+
 /**
  * struct station_info - station information
  *
@@ -2115,6 +2120,8 @@ struct station_info {
 	u8 mld_addr[ETH_ALEN] __aligned(2);
 	const u8 *assoc_resp_ies;
 	size_t assoc_resp_ies_len;
+	u16 valid_links;
+	struct link_station_info links[IEEE80211_MLD_MAX_NUM_LINKS];
 };
 
 /**
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -7030,7 +7030,8 @@ static int nl80211_send_station(struct s
 				const u8 *mac_addr, struct station_info *sinfo)
 {
 	void *hdr;
-	struct nlattr *sinfoattr, *bss_param;
+	struct nlattr *sinfoattr, *bss_param, *links, *link;
+	unsigned int link_id = 0;
 
 	hdr = nl80211hdr_put(msg, portid, seq, flags, cmd);
 	if (!hdr) {
@@ -7042,6 +7043,26 @@ static int nl80211_send_station(struct s
 	    nla_put(msg, NL80211_ATTR_MAC, ETH_ALEN, mac_addr) ||
 	    nla_put_u32(msg, NL80211_ATTR_GENERATION, sinfo->generation))
 		goto nla_put_failure;
+	if (sinfo->valid_links) {
+		links = nla_nest_start(msg,
+				       NL80211_ATTR_MLO_LINKS);
+		if (!links)
+			goto nla_put_failure;
+
+		for_each_valid_link(sinfo, link_id) {
+			link = nla_nest_start(msg, link_id + 1);
+			if (!link)
+				goto nla_put_failure;
+			if (nla_put_u8(msg, NL80211_ATTR_MLO_LINK_ID, link_id))
+				goto nla_put_failure;
+			if (nla_put(msg, NL80211_ATTR_MAC, ETH_ALEN,
+				    sinfo->links[link_id].addr))
+				goto nla_put_failure;
+			nla_nest_end(msg, link);
+		}
+
+		nla_nest_end(msg, links);
+	}
 
 	sinfoattr = nla_nest_start_noflag(msg, NL80211_ATTR_STA_INFO);
 	if (!sinfoattr)
